src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/TempDataApplicationModelProvider.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class TempDataApplicationModelProvider : IApplicationModelProvider
+    {
+        /// <inheritdoc />
+        public int Order { get { return -1000 + 10; } }
+
+        /// <inheritdoc />
+        public void OnProvidersExecuted(ApplicationModelProviderContext context)
+        {
+        }
+
+        /// <inheritdoc />
+        public void OnProvidersExecuting(ApplicationModelProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var properties = controllerModel.ControllerType.GetProperties();
+                foreach (var property in properties)
+                {
+                    if (property.CustomAttributes.Select(
+                        a => a.AttributeType == typeof(TempDataAttribute)).FirstOrDefault())
+                    {
+                        controllerModel.Filters.Add(new SaveTempDataPropertyFilterProvider());
```

Comment:
Let's not add a filter pre-property.

Instead, add one filter to the controller, and make that filter **remember** all of the property helper instances that have `[TempData]`